group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -2402,143 +2610,2090 @@ public void testOnNewMetadataImage() {
+                fromTopicPartitions(actualValue.partitionsPendingAssignment()));
+        } else {
+            assertEquals(expected.message(), actual.message());
+        }
+    }
+
+    private Map<Uuid, Set<Integer>> fromTopicPartitions(
+        List<ConsumerGroupCurrentMemberAssignmentValue.TopicPartitions> assignment
+    ) {
+        Map<Uuid, Set<Integer>> assignmentMap = new HashMap<>();
+        assignment.forEach(topicPartitions -> {
+            assignmentMap.put(topicPartitions.topicId(), new HashSet<>(topicPartitions.partitions()));
         });
         return assignmentMap;
     }
+
+    private static GenericGroup genericGroup(GroupMetadataManagerTestContext context, String groupId) {
+        return context.groupMetadataManager.getOrMaybeCreateGenericGroup(groupId, false);
+    }
+
+    private List<String> verifyGenericGroupJoinResponses(
+        List<CompletableFuture<JoinGroupResponseData>> responseFutures,
+        int expectedSuccessCount,
+        Errors expectedFailure
+    ) {
+        int successCount = 0;
+        List<String> memberIds = new ArrayList<>();
+        for (CompletableFuture<JoinGroupResponseData> responseFuture : responseFutures) {
+            if (!responseFuture.isDone()) {
+                fail("All responseFutures should be completed.");
+            }
+            try {
+                if (responseFuture.get(5, TimeUnit.SECONDS).errorCode() == Errors.NONE.code()) {
+                    successCount++;
+                } else {
+                    assertEquals(
+                        expectedFailure.code(),
+                        responseFuture.get(5, TimeUnit.SECONDS).errorCode()
+                    );
+                }
+                memberIds.add(responseFuture.get(5, TimeUnit.SECONDS).memberId());
+            } catch (Exception e) {
+                fail("Unexpected exception: " + e.getMessage());
+            }
+        }
+
+        assertEquals(expectedSuccessCount, successCount);
+        return memberIds;
+    }
+
+    /**
+     * Verify the records that should be appended and complete the append future based on a
+     * configured error. Run any assertions to verify the result of the future completion.
+     *
+     * @param expectedResult  The expected result to compare against.
+     * @param result          The result from expiring a join/heartbeat/sync operation.
+     */
+    private static void verifyCoordinatorResult(
+        ExpectedGenericGroupResult expectedResult,
+        CoordinatorResult<Void, Record> result
+    ) {
+        if (expectedResult == null) {
+            assertEquals(EMPTY_RESULT, result);
+        } else {
+            assertEquals(expectedResult.records, result.records());
+            if (expectedResult.mockError == Errors.NONE) {
+                result.appendFuture().complete(null);
+            } else {
+                result.appendFuture().completeExceptionally(expectedResult.mockError.exception());
+            }
```

Comment:
This is really surprising. `verify*` suggests that this method only verifies something but it also has side effects. I think that this should rather be done in the context like I did for the new protocol.

Reply:
removed this method, and now individual tests do the validation. 

one exception is for timer operation expirations - as the majority of the cases will not result in any records, i have done the validation inside MockCoordinatorTimer.